src/records.jl

```diff
@@ -26,49 +28,90 @@ Simply wraps the value `x` in a `Nullable` and sticks that in an `Attribute` wit
 empty `Function`.
 """
 Attribute(x::T) where {T} = Attribute{T}(() -> x)
+Attribute{T}(x::T) where {T} = Attribute{T}(() -> x)
 
 """
     get(attr::Attribute{T}) -> T
 
 Run set `attr.x` to the output of `attr.f` if `attr.x` is not already set.
 We then return the value stored in `attr.x`
 """
-function Base.get(attr::Attribute)
+function Base.get(attr::Attribute{T}) where T
     if isnull(attr.x)
-        attr.x = Nullable(attr.f())
+        attr.x = Nullable{T}(attr.f())
     end
 
-    return get(attr.x)
+    return get(attr.x)::T
 end
 
+hasfield(T::DataType, name::Symbol) = Base.fieldindex(T, name, false) > 0
+
 """
     Record
 
-Are an `Attribute` container used to store information about a log events including the
-msg, date, level, stacktrace, etc. `Formatter`s use `Records` to format log
+A dictionary-like container with `Symbol` keys used to store information about a log events
+including the msg, date, level, stacktrace, etc. `Formatter`s use `Records` to format log
 message strings.
 
-NOTE: you should access `Attribute`s in a `Record` by using `getindex` (ie: record[:msg])
-as this will correctly extract the value from the `Attribute` container.
+You can access the properties of a `Record` by using `getindex` (ie: record[:msg]).
+
+Subtypes of `Record` should implement `getindex(::MyRecord, ::Symbol)` and key-value pair
+iteration.
+"""
+abstract type Record <: AbstractDict{Symbol, Any} end
+
 """
-abstract type Record end
+    AttributeRecord <: Record
+
+A `Record` which stores its properties as `Attribute`s for lazy evaluation.
+
+Calling `getindex` or iterating will evaluate and cache the properties accessed.
+
+Subtypes of `AttributeRecord` should implement `Memento.getattribute(::MyRecord, ::Symbol)`
+instead of `getindex`.
+"""
+abstract type AttributeRecord <: Record end
```

Comment:
Couldn't we just parameterize `Record` by the value type (`AttributeRecord` could just be `Record{<:Attribute}`)?

Reply:
The problem there is the value type isn't actually `Attribute` because you want to unwrap the attributes when you pull them out.

Reply:
Okay, I may want to iterate on this hierarchy in the future, but I think this should be fine for now.